cpp/src/sampling/detail/graph_functions.hpp

```diff
@@ -120,6 +121,81 @@ sample_edges(raft::handle_t const& handle,
              size_t fanout,
              bool with_replacement);
 
+/**
+ * @brief Use the sampling results from hop N to populate the new frontier for hop N+1.
+ *
+ * @tparam vertex_t Type of vertex identifiers. Needs to be an integral type.
+ * @tparam label_t Type of label. Needs to be an integral type.
+ * @tparam multi_gpu Flag indicating whether template instantiation should target single-GPU (false)
+ *
+ * @param handle RAFT handle object to encapsulate resources (e.g. CUDA stream, communicator, and
+ * handles to various CUDA libraries) to run graph algorithms.
+ * @param starting_vertices The source vertices for the current hop
+ * @param starting_vertex_labels Optional labels for the source vertices for the current hop
+ * @param new_frontier_vertices Vertices discovered in the current hop
+ * @param new_frontier_vertex_labels Optional labels for each of the newly discovered vertices
```

Comment:
Better rename these variables to `next_frontier_vertices` & `next_frontier_vertex_labels`? (to be consistent with the function name?). And conceptually, we are taking the candidates for the next frontier; modifying them; and returning the modified candidates. I think it is more intuitive to take these as R-values (`vertex_used_as_source` as `raft::device_span<T const>`).

Reply:
I think we can achieve this if we require `vertex_used_as_source` to be sorted and use `set_union` for actual implementation.

Reply:
renamed in next push

Note that `vertex_used_as_source` is modified in this function.  This is why it is moved in and returned as a result.  I could modify this to be:

```
  std::optional<std::tuple<rmm::device_uvector<vertex_t>,
                           std::optional<rmm::device_uvector<label_t>>>>& vertex_used_as_source,
```

and remove it from the return value if you think that's more intuitive for this use case.

Reply:
I will look at the `set_union` option next.